### ISSUE TITLE
Incorrect `not` behavior

### ISSUE THREAD

**Issue by seldridge** [freechipsproject/firrtl-interpreter#122]
This showed up in https://github.com/freechipsproject/firrtl/issues/753.

Consider the following test case:
```scala
// Just adding to `ConcreteSpec`
class ConcreteSpec extends FlatSpec with Matchers {

  // rest of the spec...

  behavior of "firrtl:issue-753"
  it should "invert an SInt" in {
    val a = ConcreteSInt(-1, 32)
    val notA = a.not
    val notNotA = notA.not
    val signedNotNotA = notNotA.asSInt
    println("a:            " + a.value.toString(16))
    println("~a:           " + notA.value.toString(16))
    println("~(~a):        " + notNotA.value.toString(16))
    println("$signed~(~a): " + signedNotNotA.value.toString(16))
    (signedNotNotA.value) should be (a.value)
  }
}
```

This fails and the println output is:
```
a:            -1
~a:           7ffffffe
~(~a):        80000001
$signed~(~a): -7fffffff
```

I believe that this disagrees with the FIRRTL specification for `not`. Specifically, "The bitwise complement operation performs a logical not on each bit in e," returning a `UInt` in either case. If I'm interpreting that correctly, then `not(-1)` should be `0`.

A fix would involve some minor modifications in Concrete here: https://github.com/freechipsproject/firrtl-interpreter/blob/master/src/main/scala/firrtl_interpreter/Concrete.scala#L185.

**Comment by lfiolhais** [comment]
I believe I figured out the bug. The code @seldridge linked ignores two's complement representation by separating the number into two components, its absolute value and its sign. Thus, transforming it into a sign-magnitude representation.

The fix seems fairly simple to me (I'm no Scala expert so please correct me if I'm wrong):
```scala
def not: ConcreteUInt = this match {
    case ConcreteUInt(v, _, p) | ConcreteSInt(v, _, p) =>
      var flipped = value
      for(bitIndex <- 0 until width) {
        flipped = flipped.flipBit(bitIndex)
      }
      ConcreteUInt(flipped, width, p)
}
```
In this case the function will invert every bit in the number, regardless of sign.

If this is correct I'd be more than happy to submit a PR.

**Comment by seldridge** [comment]
I think the fix is a little more complicated, but the general approach seems right. You want to do a two's complement conversion on the `value: BigInt` then flip all its bits.

If you want to play with this, there are a bunch of existing `ConcreteSpec` tests that you can run. Startup `sbt` in the `firrtl-interpreter` top level and then run `testOnly firrtl_interpreter.ConcreteSpec`. You can add the small test that I have for your specific case in there temporarily to make sure that passes. The fixes that I tried aren't passing "flip bits of UInts", but I think that test is wonky. Similarly, I'm not convinced that `asBinaryString` is doing the right thing...

**Comment by chick** [comment]
It is a bit more complicated, and furthermore I think there is a problem with 
SInt#toBinaryString.  I'm digging into it.
Thanks for all the help.

**Comment by seldridge** [comment]
Thanks for digging into this @chick!

**Comment by lfiolhais** [comment]
Thanks for helping out everyone! In the mean time if someone else falls into the same problem the workaround is to cast the signal to UInt, perform the complement, and then cast it back to SInt.
```scala
val x = 5.S
val y = (~(x.asUInt)).asSInt
```

**Comment by chick** [comment]
@seldridge @lfiolhais I have created a PR #123 
Take a look. I re-did the not test to be a bit more comprehensible.
It may seem a bit tautological now, but I think since not does it's work in a different way the
test should be good enough

**Comment by seldridge** [comment]
Fixed by #123
